dlt/destinations/path_utils.py

```diff
@@ -1,45 +1,225 @@
+                        " parameters below`schema name`, `table name`, `load_id`, `file_id`,"
+                        " `extension` and `current_datetime`",
+                        fg="red",
+                    )
+                    raise
+            else:
+                params[key] = value
+
+    return params
+
+
+def check_layout(
+    layout: str,
+    extra_placeholders: Optional[Dict[str, Any]] = None,
+    allowed_placeholders: Optional[Set[str]] = SUPPORTED_PLACEHOLDERS,
+) -> Tuple[List[str], List[str]]:
+    """Returns a tuple with all valid placeholders and the list of layout placeholders
+
+    Raises: InvalidFilesystemLayout
+
+    Returns: a pair of lists of valid placeholders and layout placeholders
+    """
     placeholders = get_placeholders(layout)
-    invalid_placeholders = [p for p in placeholders if p not in SUPPORTED_PLACEHOLDERS]
+    # Build out the list of placeholder names
+    # which we will use to validate placeholders
+    # in a given config.layout template
+    all_placeholders = allowed_placeholders.copy()
+    if extra_placeholders:
+        for placeholder, _ in extra_placeholders.items():
+            all_placeholders.add(placeholder)
+
+    # now collect all unknown placeholders from config.layout template
+    invalid_placeholders = [p for p in placeholders if p not in all_placeholders]
     if invalid_placeholders:
         raise InvalidFilesystemLayout(invalid_placeholders)
-    return placeholders
 
-
-def get_placeholders(layout: str) -> List[str]:
-    return re.findall(r"\{(.*?)\}", layout)
+    return list(all_placeholders), placeholders
 
 
 def create_path(
-    layout: str, schema_name: str, table_name: str, load_id: str, file_id: str, ext: str
+    layout: str,
+    file_name: str,
+    schema_name: str,
+    load_id: str,
+    load_package_timestamp: Optional[str] = None,
+    current_datetime: Optional[TCurrentDateTime] = None,
+    extra_placeholders: Optional[Dict[str, Any]] = None,
 ) -> str:
     """create a filepath from the layout and our default params"""
-    placeholders = check_layout(layout)
-    path = layout.format(
+    if callable(current_datetime):
+        current_datetime = current_datetime()
+
+    if current_datetime and not isinstance(current_datetime, pendulum.DateTime):
+        raise RuntimeError("current_datetime is not an instance instance of pendulum.DateTime")
+
+    job_info = ParsedLoadJobFileName.parse(file_name)
+    params = prepare_params(
+        extra_placeholders=extra_placeholders,
+        job_info=job_info,
         schema_name=schema_name,
-        table_name=table_name,
         load_id=load_id,
-        file_id=file_id,
-        ext=ext,
-        curr_date=str(pendulum.today()),
     )
+
+    datetime_params = prepare_datetime_params(current_datetime, load_package_timestamp)
+    params.update(datetime_params)
+
+    placeholders, layout_placeholders = check_layout(layout, params)
```

Comment:
again: this is way too late. you can validate layout in the factory class for filesystem, just after configuration is resolved. (override `configuration` method). at this point you'll have all the arguments merged

Reply:
I will lift this up into configuration resolve time then.

Reply:
Done.